distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso

```diff
@@ -0,0 +1,47 @@
+from Standard.Base import all
+import Standard.Base.Data.Read.Many_Files_List.Many_Files_List
+import Standard.Base.Data.Read.Return_As.Return_As
+import Standard.Base.Errors.Common.Type_Error
+from Standard.Base.Metadata.Choice import Option
+
+import project.Columns_To_Keep.Columns_To_Keep
+import project.Internal.Read_Many_Helpers
+import project.Match_Columns.Match_Columns
+
+type Return_As_Table
+    ## Returns a table with a new column `Value` containing the objects loaded
+       from each file.
+
+       When the source for files to load was a table, all columns from the
+       original table are also retained. In case of name clashes, the newly
+       added columns will get a suffix.
+
+       When the source was a simple Vector, the returned table will also contain
+       a `File Name` column.
+    Table_Of_Objects
```

Comment:
The name `Table_Of_Objects` was a working name. I'm not sure if we want to keep it or come up with something better?

Definitely `Table_Of_Tables` in Excel sounded much better. But here it may not necessarily be tables, it may be various objects. So `Table_Of_Objects` is quite precise, but it sounds weird.

Reply:
Maybe we should change approach altogether and try something like `Added_Column`? (meaning that the loaded files are added as a column to the input)

- `Values_Column`
- `Table_With_New_Column`

any thoughts on these names?

Reply:
I think `Table_With_New_Column` or `Table_Of_Objects` seem the best to me, `Values_Column` I think could end up being confusing

Reply:
I wonder if:

```
type Return_As
   ## Single table with values expanded
   Merged_Table

   ## PRIVATE
       Add this for backwards compatiblity then hide in widget?
   Table_Of_Tables

   ## 
   Individual_Cells

   ## Objects returned as a raw vector allowing zipping to input
   Vector
```

Not sure on name yet...

Reply:
> How about more of a boolean? Merged/UnMerged. Merged True/False?

But note that `Merged_Table` has additional arguments (`columns_to_keep` and `match`) that are only relevant in this mode. Switching to a boolean would make this problematic.

Reply:
> I wonder if:
> 
> ```
> type Return_As
>    ## Single table with values expanded
>    Merged_Table
> 
>    ## PRIVATE
>        Add this for backwards compatiblity then hide in widget?
>    Table_Of_Tables
> 
>    ## 
>    Individual_Cells
> 
>    ## Objects returned as a raw vector allowing zipping to input
>    Vector
> ```
> 
> Not sure on name yet either - will think some more

`Vector` is handled by `Return_As_Base` - the type needs to be split up so that the `Vector` return works even if `Table` is not imported. It is opaque to the user - the dropdown displays all available types and autoscoping works as shown in tests.

I'd not add `Table_Of_Tables` to `Data.read_many`. It is supported in `Excel_Workbook.read_many` to retain backwards compatibility. But adding it to a brand new function seems unnecessary - why add an obsolete option?

Reply:
> > How about more of a boolean? Merged/UnMerged. Merged True/False?
> 
> But note that `Merged_Table` has additional arguments (`columns_to_keep` and `match`) that are only relevant in this mode. Switching to a boolean would make this problematic.

Plus dropdown options are clearer in the GUI generally.

Reply:
Yes sorry that wasn't clear. When I said boolean I didn't mean a boolean type. More text options that were Merged or Not Merged.

Reply:
> `Vector` is handled by `Return_As_Base` - the type needs to be split up so that the `Vector` return works even if `Table` is not imported. It is opaque to the user - the dropdown displays all available types and autoscoping works as shown in tests.
> 
> I'd not add `Table_Of_Tables` to `Data.read_many`. It is supported in `Excel_Workbook.read_many` to retain backwards compatibility. But adding it to a brand new function seems unnecessary - why add an obsolete option?

I was in my head thinking of it as a single type used by both functions but I appreciate we are actually building this dynamically. Just for working out how it might look in the dropdown thinking as a single type is a bit easier!

If the `..Table_Of_Tables` was an option not shown in the dropdown but in one of the *mix in* types then no special handling would be needed for the two reads to share the same object was my thought.

I wonder if:
- `..As_Vector`
- `..With_New_Column` 
- `..As_Merged_Table`

Might be cleanest?

Reply:
> If the `..Table_Of_Tables` was an option not shown in the dropdown but in one of the _mix in_ types then no special handling would be needed for the two reads to share the same object was my thought.

But then more code would be affected by the legacy options, I'd rather not grow places that take legacy options when not needed. Unless there's any problem with the current approach, I'd keep the compatibility conversion as-is.



> I wonder if:
> 
> * `..As_Vector`
> * `..With_New_Column`
> * `..As_Merged_Table`
> 
> Might be cleanest?

That sounds quite clear to me 👍

Reply:
Yes I like those options

Reply:
The options outlined in the comment above have been implemented. `Excel_Workbook.read_many` also allows old options for compatibility, but `Data.read_many` (the new component) only works with new ones.